smithy-model/src/main/java/software/amazon/smithy/model/knowledge/TextInstance.java

```diff
@@ -0,0 +1,120 @@
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.knowledge;
+
+import java.util.Deque;
+import java.util.List;
+import java.util.Objects;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.utils.ListUtils;
+
+/**
+ * Contains information about text that occurs in the Smithy IDL written by the owner,
+ * inluding location metadata.
+ */
+public final class TextInstance {
+    private final TextLocationType locationType;
+    private final String text;
+    private final Shape shape;
+    private final Trait trait;
+    private final List<String> traitPropertyPath;
+
+    private TextInstance(
+            final TextLocationType locationType,
+            final String text,
+            final Shape shape,
+            final Trait trait,
+            final Deque<String> traitPropertyPath
+    ) {
+        this.locationType = locationType;
+        this.text = text;
+        this.shape = shape;
+        this.trait = trait;
+        this.traitPropertyPath = traitPropertyPath != null
+            ? ListUtils.copyOf(traitPropertyPath)
+            : null;
+    }
+
+    static TextInstance createNamespaceText(String namespace) {
+        Objects.requireNonNull(namespace, "'namespace' must be specified");
+        return new TextInstance(TextLocationType.NAMESPACE, namespace, null, null, null);
+    }
+
+    static TextInstance createShapeInstance(Shape shape) {
+        Objects.requireNonNull(shape, "'shape' must be specified");
+        return new TextInstance(TextLocationType.SHAPE, shape.getId()
+                    .getMember().orElseGet(() -> shape.getId().getName()),
+                    shape, null, null);
+    }
+
+    static TextInstance createTraitInstance(String text, Shape shape, Trait trait, Deque<String> traitPath) {
+        Objects.requireNonNull(trait, "'trait' must be specified");
+        Objects.requireNonNull(shape, "'shape' must be specified");
+        Objects.requireNonNull(text, "'text' must be specified");
+        return new TextInstance(TextLocationType.APPLIED_TRAIT, text, shape, trait, traitPath);
+    }
+
+    /**
+     * Retrieves the type of TextLocationType associated with the text.
+     */
+    public TextLocationType getLocationType() {
+        return locationType;
+    }
+
+    /**
+     * Retrieves the text content of the TextInstance.
+     */
+    public String getText() {
+        return text;
+    }
+
+    /**
+     * Gets the shape the text is associated with.
+     *
+     * Returns null if {@link TextLocationType} is NAMESPACE
```

Comment:
Instead of returning `null`, we almost always use `Optional<Shape>` for the return type and use `Optional.ofNullable()` when returning. Applies to `getTrait` as well.

Also, the javadoc should use `@return` to describe return values so they're properly integrated with the generated documentation.